README.md
# dtrack

dtrack 是 DependencyTrack 的cli程序

```
Usage: dtrack [OPTIONS]

Options:
  -u, --url <Url>                  设置 Dependency-track 的访问地址. eg. http://dtrack.abc.local:8081 [default: http://127.0.0.1:8081]
  -k, --key <Apikey>               设置 Dependency-track 的apikey. eg. adfadfe343g [default: Oh9LHLfrLgk77e67DEZtiitOWZwvFVXI]
      --proxy <Proxy>              设置 Dependency-track 的连接代理. eg. http://127.0.0.1:8080
  -p, --project <Project Name>     设置 Dependency-track 的项目名称 [default: test]
  -e, --edition <Project Version>  设置 Dependency-track 的项目版本 [default: default]
  -f, --file <Bom File>            设置 Dependency-track 的文件
  -r, --rule <Rule>                设置 Dependency-track 的规则
  -d, --dversion                   输出 Dependency-track 版本信息
  -o, --output <Scan Result>       设置 Dependency-track 的扫描结果存储位置 [default: results.json]
  -l, --log-level <LOG_LEVEL>      设置输出日志的级别(选择off不输出日志) [default: debug] [possible values: off, debug, info, warn, error]
  -h, --help                       Print help
  -V, --version                    Print version
```

## 如何下载

### 1.Cargo
```
cargo install dtrack
```
### 2.Windows
```
直接下载dtrack-x86_64-pc-windows-msvc.zip解压即可
```

### 3.Linux
```
LATEST_VERSION=$(curl -s https://api.github.com/repos/cnsino/dtrack/releases/latest | jq -r '.tag_name')
DOWNLOAD_URL="https://github.com/cnsino/dtrack/releases/download/$LATEST_VERSION/dtrack-x86_64-unknown-linux-gnu.tar.gz"
curl -L $DOWNLOAD_URL
tar -xzvf dtrack-x86_64-unknown-linux-gnu.tar.gz
mv dtrack /usr/local/bin/
```

## 使用方式
```bash
dtrack -p test2 -e 0.0.1 -f /root/bom.xml -o /root/results.json
```

CHANGELOG.md
# Changelog

## [Unreleased]

### Added
- New feature A
- New feature B

### Changed
- Improved feature C

## [1.0.0] - 2023-06-30

### Added
- 添加--proxy,--log-level参数
- 处理错误
- 发布完整版本

## [0.0.2] - 2023-06-25

### Init
- 发布可用版本


src/lib.rs
use clap::Parser;
use error::DtrackError;
use std::process;

mod args;
mod auditor;
pub mod error;
pub async fn builder() -> Result<(), DtrackError> {
    let args = args::Dtrack::parse();
    env_logger::Builder::new()
        .filter_level(log::LevelFilter::Off)
        .filter_module("dtrack", args.log_level)
        .init();

    let auditor = auditor::Auditor::new(args.url.unwrap(), args.proxy, args.key.unwrap()).await?;
    log::debug!("{auditor:?}");
    if args.dversion {
        if let Some(version) = auditor.version {
            println!("当前 Dependency-Track的安装版本是：{}", version);
        } else {
            println!("未能获取到Dependency-Track的安装版本");
        }
        process::exit(0);
    }

    if let (Some(project), Some(project_version), Some(bom_file), Some(output)) =
        (&args.project, &args.edition, &args.file, &args.output)
    {
        log::debug!("输入的参数：项目：{project}, 版本：{project_version}, Bom文件：{bom_file}, 输出文件：{output}");
        let token = Some(
            auditor
                .read_upload_bom(project, project_version, bom_file)
                .await?,
        );
        if let Some(token) = token {
            log::info!("Bom: {:?}扫描开始", token);
            let scan_status = auditor
                .poll_bom_token_being_processed(token.as_str())
                .await?;
            if scan_status {
                log::info!("Bom: {:?}扫描完成", token)
            }
        }
        let project_uuid = auditor.get_project_uuid(project, project_version).await?;
        log::info!("获取的项目id{:?}", project_uuid);
        auditor
            .get_project_findings(project_uuid.as_str(), output)
            .await?;
    } else {
        log::error!("{}", DtrackError::ParamMissing);
        return Err(DtrackError::ParamMissing);
    }
    Ok(())
}


src/auditor.rs
use std::{
    fmt::Display,
    fs::{self, File},
    io::{Read, Write},
    path::Path,
    process,
    time::Duration,
};

use crate::error::DtrackError;
use reqwest::{header, multipart, Client, IntoUrl, StatusCode};
use serde_json::Value;
#[derive(Debug, Clone)]
pub(crate) struct Auditor<T> {
    url: T,
    client: Client,
    pub(crate) version: Option<String>,
}

impl<T> Auditor<T>
where
    T: IntoUrl + Display,
{
    pub(crate) async fn new<U>(url: T, proxy: Option<U>, key: String) -> Result<Self, DtrackError>
    where
        U: IntoUrl + Display,
    {
        let mut headers = header::HeaderMap::new();
        headers.insert(
            header::HeaderName::from_static("x-api-key"),
            header::HeaderValue::from_str(&key).map_err(|_| DtrackError::InvalidHeader(key))?,
        );
        log::debug!("构造的请求头：{headers:?}");
        let mut builder = Client::builder();
        if let Some(proxy) = proxy {
            log::debug!("使用的代理：{proxy}");
            let proxy = reqwest::Proxy::all(proxy).map_err(|_| DtrackError::ProxyBuildError)?;
            builder = builder.proxy(proxy);
        }

        let client = builder
            .http1_title_case_headers()
            .default_headers(headers)
            .danger_accept_invalid_certs(true)
            .build()
            .map_err(|_| DtrackError::ClientBuildError)?;

        let mut auditor = Self {
            url,
            client,
            version: None,
        };
        match auditor.set_version().await {
            Ok(()) => log::debug!("Version 获取结果: {}", auditor.version.is_some()),
            Err(e) => {
                log::error!("Error: {}", e);
                eprintln!("程序运行出错: {}", e);
                process::exit(1);
            }
        };
        Ok(auditor)
    }

    async fn set_version(&mut self) -> Result<(), DtrackError> {
        let version_url = format!("{}api/version", self.url);
        log::debug!("version_url{:?}", version_url);
        let response = self
            .client
            .get(&version_url)
            .send()
            .await
            .map_err(|_| DtrackError::UnableToConnect(version_url))?;

        if response.status().is_success() {
            let res = response
                .json::<Value>()
                .await
                .map_err(|_| DtrackError::UnableToParseJson)?;
            if let Some(version) = res["version"].as_str() {
                log::info!("连接服务器成功!");
                log::debug!("当前版本：{:?}", version);
                self.version = Some(version.to_string());
                return Ok(());
            } else {
                return Err(DtrackError::CustomInvalidInfo(
                    "无法从响应获取版本信息".to_owned()
                ));
            }
        } else {
            log::debug!("{}", response.status());
            log::debug!(
                "{:?}",
                response
                    .text()
                    .await
                    .map_err(|_| DtrackError::UnableToGetText)?
            );
            log::error!("连接服务器失败,请检查URL地址及key值!");
        }
        Err(DtrackError::CustomInvalidInfo(
            "连接服务器失败,请检查URL地址及key值".to_owned(),
        ))
    }

    pub(crate) async fn read_upload_bom(
        &self,
        project: &str,
        project_version: &str,
        bom_file: &str,
    ) -> Result<String, DtrackError> {
        let upload_url = format!("{}api/v1/bom", self.url);
        log::debug!("upload_url{:?}", upload_url);
        let file_path = Path::new(bom_file);
        let mut file = fs::File::open(file_path)?;
        let metadata = fs::metadata(file_path)?;
        let mut buffer = Vec::new();
        file.read_to_end(&mut buffer)?;

        let form = multipart::Form::new()
            .text("autoCreate", "true")
            .text("projectName", project.to_string())
            .text("projectVersion", project_version.to_string())
            .part(
                "bom",
                multipart::Part::stream_with_length(buffer, metadata.len()).file_name(
                    file_path
                        .file_name()
                        .unwrap()
                        .to_string_lossy()
                        .into_owned(),
                ),
            );

        log::debug!("{:?}", form);
        let response = self
            .client
            .post(&upload_url)
            .multipart(form)
            .send()
            .await
            .map_err(|_| DtrackError::UnableToConnect(upload_url))?;
        if response.status().is_success() {
            log::info!("上传bom文件成功!");
            let results = response
                .json::<Value>()
                .await
                .map_err(|_| DtrackError::UnableToParseJson)?;
            log::info!("从服务端获取项目id完成!");
            if let Some(token) = results["token"].as_str() {
                return Ok(token.to_string());
            }
        } else {
            log::debug!("{}", response.status());
            log::debug!(
                "{:?}",
                response
                    .text()
                    .await
                    .map_err(|_| DtrackError::UnableToGetText)?
            );
            log::error!("上传bom文件失败,请检查URL地址及key值!");
        }
        Err(DtrackError::CustomInvalidInfo(
            "上传bom文件失败,请检查URL地址及key值".to_owned(),
        ))
    }

    pub(crate) async fn poll_bom_token_being_processed(
        &self,
        token: &str,
    ) -> Result<bool, DtrackError> {
        let token_url = format!("{}api/v1/bom/token/{}", self.url, token);
        loop {
            log::debug!("token_url{:?}", token_url);
            let response = self
                .client
                .get(&token_url)
                .send()
                .await
                .map_err(|_| DtrackError::UnableToConnect(token_url.to_string()))?
                .json::<Value>()
                .await
                .map_err(|_| DtrackError::UnableToParseJson)?;
            if let Some(process) = response["processing"].as_bool() {
                log::info!("SCA扫描正在处理中: {:?}", process);
                if !process {
                    break;
                }
            }
            tokio::time::sleep(Duration::from_secs(5)).await;
        }
        Ok(true)
    }

    pub(crate) async fn get_project_uuid(
        &self,
        project: &str,
        project_version: &str,
    ) -> Result<String, DtrackError> {
        let project_url = format!(
            "{}api/v1/project/lookup?name={}&version={}",
            self.url, project, project_version
        );
        log::debug!("project_url{:?}", project_url);
        let response = self
            .client
            .get(&project_url)
            .send()
            .await
            .map_err(|_| DtrackError::UnableToConnect(project_url))?;
        if response.status().is_success() {
            // 解析响应体为 JSON
            let results = response
                .json::<Value>()
                .await
                .map_err(|_| DtrackError::UnableToParseJson)?;
            // 将 JSON 写入文件
            log::info!("从服务端获取项目id完成!");
            if let Some(uuid) = results["uuid"].as_str() {
                return Ok(uuid.to_string());
            }
        } else {
            log::error!("无法从服务端获取项目,请检查URL地址及key值.");
        }
        Err(DtrackError::CustomInvalidInfo(
            "无法从服务端获取项目,请检查URL地址及key值".to_owned(),
        ))
    }

    pub(crate) async fn get_project_findings(
        &self,
        project_uuid: &str,
        output: &str,
    ) -> Result<(), DtrackError> {
        let findings_url = format!("{}api/v1/finding/project/{}/export", self.url, project_uuid);
        log::debug!("findings_url{:?}", findings_url);
        let response = self
            .client
            .get(&findings_url)
            .send()
            .await
            .map_err(|_| DtrackError::UnableToConnect(findings_url))?;

        if response.status().is_success() {
            // 解析响应体为 JSON
            let results = response
                .text()
                .await
                .map_err(|_| DtrackError::UnableToGetText)?;
            // 将 JSON 写入文件
            let mut file = File::create(output)?;
            log::debug!("{:?}", file.metadata());
            file.write_all(results.as_bytes())?;
            log::info!("将扫描结果写入文件{:?}完成!", output);
            Ok(())
        } else if response.status() == StatusCode::NOT_FOUND {
            log::error!("项目名称及版本不存在");
            Err(DtrackError::CustomInvalidInfo(
                "项目名称及版本不存在".to_owned(),
            ))
        } else {
            log::error!("无法从服务端获取扫描结果,请检查URL地址及key值.");
            Err(DtrackError::CustomInvalidInfo(
                "无法从服务端获取扫描结果,请检查URL地址及key值".to_owned(),
            ))
        }
    }
}


src/args.rs
use std::str::FromStr;

use clap::builder::TypedValueParser as _;
use clap::Parser;
use reqwest::Url;
/// Dependency-track 的命令行程序
#[derive(Parser, Debug, Clone)]
#[command(author,version, about, long_about = None)]
pub(crate) struct Dtrack {
    /// 设置 Dependency-track 的访问地址. eg. http://dtrack.abc.local:8081.
    #[arg(
        short,
        long,
        value_name = "Url",
        default_value = "http://127.0.0.1:8081"
    )]
    pub(crate) url: Option<Url>,
    /// 设置 Dependency-track 的apikey. eg. adfadfe343g.
    #[arg(
        short,
        long,
        value_name = "Apikey",
        default_value = "Oh9LHLfrLgk77e67DEZtiitOWZwvFVXI"
    )]
    pub(crate) key: Option<String>,

    /// 设置 Dependency-track 的连接代理. eg. http://127.0.0.1:8080.
    #[arg(
        long,
        value_name = "Proxy",
    )]
    pub(crate) proxy: Option<Url>,
    /// 设置 Dependency-track 的项目名称.
    #[arg(short, long, value_name = "Project Name", default_value = "test")]
    pub(crate) project: Option<String>,
    /// 设置 Dependency-track 的项目版本.
    #[arg(short, long, value_name = "Project Version", default_value = "default")]
    pub(crate) edition: Option<String>,
    /// 设置 Dependency-track 的文件.
    #[arg(short, long, value_name = "Bom File")]
    pub(crate) file: Option<String>,
    /// 设置 Dependency-track 的规则.
    #[arg(short, long, value_name = "Rule")]
    pub(crate) rule: Option<String>,
    /// 输出 Dependency-track 版本信息.
    #[arg(short, long, default_value_t = false)]
    pub(crate) dversion: bool,
    /// 设置 Dependency-track 的扫描结果存储位置.
    #[arg(
        short,
        long,
        value_name = "Scan Result",
        default_value = "results.json"
    )]
    pub(crate) output: Option<String>,
    /// 设置输出日志的级别(选择off不输出日志)
    #[arg(
        short,
        long,
        default_value = "debug",
        value_parser = clap::builder::PossibleValuesParser::new(["off", "debug", "info", "warn", "error"])
            .map(|s| log::LevelFilter::from_str(&s).unwrap()),
    )]
    pub(crate) log_level: log::LevelFilter,
}
